# 📘 Day 25
## Pandas

Pandas is an open source,high-performance, easy-to-use data structures and data analysis tools for the Python programming language.
Pandas adds data structures and tools designed to work with table-like data which is Series and Data Frames
Pandas provides tools for data manipulation: reshaping, merging, sorting, slicing, aggregation and imputation.
```py
pip install conda
conda install pandas
```
Pandas data structure is based on *Series* and *DataFrames*
A series is a column and a DataFrame is a multidimensional table made up of collection of series. In order to create a pandas series we should use numpy to create a one dimensional arrays or a python list.
Let's see an example of a series:

Names pandas Series

![pandas series](images/pandas-series-1.png) 

Countries Series

![pandas series](images/pandas-series-2.png) 

Cities Series

![pandas series](images/pandas-series-3.png)

As you can see, pandas series is just one column data. If we want to have multiple columns we use data frames. The example below shows pandas DataFrames.

Let's see, an example of a pandas data frame:

![Pandas data frame](images/pandas-dataframe-1.png)

Data from is a collection of rows and columns. Look at the table below it has many columns than the above


![Pandas data frame](images/pandas-dataframe-2.png)

Next, we will see how to import pandas and how to create Series and DataFrames using pandas

## Importing pandas

In [359]:
import pandas as pd # importing pandas as pd
import numpy  as np # importing numpy as np

### Creating Pandas Series with default index

In [360]:
nums = [1, 2, 3, 4,5]
s = pd.Series(nums)
s

0    1
1    2
2    3
3    4
4    5
dtype: int64

### Creating  Pandas Series with custom index

In [361]:
nums = [1, 2, 3, 4, 5]
s = pd.Series(nums, index=[1, 2, 3, 4, 5])
s


1    1
2    2
3    3
4    4
5    5
dtype: int64

In [362]:
fruits = ['Orange','Banana','Mangao']
fruits = pd.Series(fruits, index=[1, 2, 3])
fruits

1    Orange
2    Banana
3    Mangao
dtype: object

### Creating Pandas Series from a dictionary

In [363]:
dct = {'name':'Asabeneh','country':'Finland','city':'Helsinki'}

In [364]:
s = pd.Series(dct)
s

name       Asabeneh
country     Finland
city       Helsinki
dtype: object

### Creating a constant pandas series

In [365]:
s = pd.Series(10, index = [1, 2,3])
s

1    10
2    10
3    10
dtype: int64

### Creating a  pandas series using linspace

In [366]:
s = pd.Series(np.linspace(5, 20, 10)) # linspace(starting, end, items)
s

0     5.000000
1     6.666667
2     8.333333
3    10.000000
4    11.666667
5    13.333333
6    15.000000
7    16.666667
8    18.333333
9    20.000000
dtype: float64

## DataFrames

Pandas data frames can be created in different ways.

### Creating DataFrames from list of lists

In [367]:
data = [
    ['Asabeneh', 'Finland', 'Helsink'], 
    ['David', 'UK', 'London'],
    ['John', 'Sweden', 'Stockholm']
]
df = pd.DataFrame(data, columns=['Names','Country','City'])
df


Names  Country       City
0  Asabeneh  Finland    Helsink
1     David       UK     London
2      John   Sweden  Stockholm

### Creating DataFrame using Dictionary

In [368]:
data = {'Name': ['Asabeneh', 'David', 'John'], 'Country':[
    'Finland', 'UK', 'Sweden'], 'City': ['Helsiki', 'London', 'Stockholm']}
df = pd.DataFrame(data)
df

Name  Country       City
0  Asabeneh  Finland    Helsiki
1     David       UK     London
2      John   Sweden  Stockholm

### Creating DataFrams from list of dictionaries

In [369]:
data = [
    {'Name': 'Asabeneh', 'Country': 'Finland', 'City': 'Helsinki'},
    {'Name': 'David', 'Country': 'UK', 'City': 'London'},
    {'Name': 'John', 'Country': 'Sweden', 'City': 'Stockholm'}]
df = pd.DataFrame(data)
df

Name  Country       City
0  Asabeneh  Finland   Helsinki
1     David       UK     London
2      John   Sweden  Stockholm

## Reading CSV File using pandas

In [370]:
import pandas as pd

df = pd.read_csv('./data/weight-height.csv')

### Data Exploration
Let's read only the first 5 rows using head()

In [371]:
df.head() # give five rows we can increase the number of rows by passing argument to the head() method

Gender     Height      Weight
0   Male  73.847017  241.893563
1   Male  68.781904  162.310473
2   Male  74.110105  212.740856
3   Male  71.730978  220.042470
4   Male  69.881796  206.349801

As you can see the csv file has three rows:Gender, Height and Weight. But we don't know the number of rows. Let's use shape meathod.

In [372]:
df.shape # as you can see 10000 rows and three columns

(10000, 3)

Let's get all the columns using columns.


In [373]:
df.columns

Index(['Gender', 'Height', 'Weight'], dtype='object')

Let's read only the last 5 rows using tail()

In [374]:
df.tail() # tails give the last five rows, we can increase the rows by passing argument to tail method

Gender     Height      Weight
9995  Female  66.172652  136.777454
9996  Female  67.067155  170.867906
9997  Female  63.867992  128.475319
9998  Female  69.034243  163.852461
9999  Female  61.944246  113.649103

Now, lets get specif colums using the column key


In [375]:
heights = df['Height'] # this is now a a series

In [376]:
heights

0       73.847017
1       68.781904
2       74.110105
3       71.730978
4       69.881796
          ...    
9995    66.172652
9996    67.067155
9997    63.867992
9998    69.034243
9999    61.944246
Name: Height, Length: 10000, dtype: float64

In [377]:
weights = df['Weight'] # this is now a series

In [378]:
weights

0       241.893563
1       162.310473
2       212.740856
3       220.042470
4       206.349801
           ...    
9995    136.777454
9996    170.867906
9997    128.475319
9998    163.852461
9999    113.649103
Name: Weight, Length: 10000, dtype: float64

In [379]:
len(heights) == len(weights)

True

In [380]:
heights.describe() # give statisical information about height data

count    10000.000000
mean        66.367560
std          3.847528
min         54.263133
25%         63.505620
50%         66.318070
75%         69.174262
max         78.998742
Name: Height, dtype: float64

In [381]:
weights.describe()

count    10000.000000
mean       161.440357
std         32.108439
min         64.700127
25%        135.818051
50%        161.212928
75%        187.169525
max        269.989699
Name: Weight, dtype: float64

In [382]:
df.describe()  # describe can also give statistical information from a datafrom

Height        Weight
count  10000.000000  10000.000000
mean      66.367560    161.440357
std        3.847528     32.108439
min       54.263133     64.700127
25%       63.505620    135.818051
50%       66.318070    161.212928
75%       69.174262    187.169525
max       78.998742    269.989699

## Modifying DataFrame



Modifying a DataFrame
    * We can create a new DataFrame
    * We can create a new column and add to DataFrame, 
    * we can remove an existing column from DataFrame, 
    * we can modify an existing column from DataFrame, 
    * we can change the data type of column values from DataFrame

### Create a DataFrame
All the time, first we import the necessary packages. Now, lets import pandas and numpy two best friends ever.

In [383]:
import pandas as pd
import numpy as np
data = [
    {"Name": "Asabeneh", "Country":"Finland","City":"Helsinki"},
    {"Name": "David", "Country":"UK","City":"London"},
    {"Name": "John", "Country":"Sweden","City":"Stockholm"}]
df = pd.DataFrame(data)
df

Name  Country       City
0  Asabeneh  Finland   Helsinki
1     David       UK     London
2      John   Sweden  Stockholm

Adding column in DataFrame is like adding a key in dictionary.

First let's use the previous example to create a DataFrame. After we create the DataFrame,  we will start modifying the columns and column values.

### Adding new column
Let's add a weight column in the DataFrame

In [384]:
weights = [74, 78, 69]
df['Weight'] = weights
df

Name  Country       City  Weight
0  Asabeneh  Finland   Helsinki      74
1     David       UK     London      78
2      John   Sweden  Stockholm      69

Let's add a height column in the DataFrame

In [385]:
heights = [173, 175, 169]
df['Height'] =heights
df

Name  Country       City  Weight  Height
0  Asabeneh  Finland   Helsinki      74     173
1     David       UK     London      78     175
2      John   Sweden  Stockholm      69     169

As you can see from the above DataFrame, now we new added columns, the Weight and Height. Let's add one additional column by called BMI(Body Mass Index) by calculating their BMI using thier mass and height. BMI is mass divided by height square meter(Weight/Height * Height).

As you can see, the hieght is in centimeter, so we shoud change the height to meter. So, let's modify the height row

### Modifying column values

In [386]:
df['Height'] = df['Height'] * 0.01
df


Name  Country       City  Weight  Height
0  Asabeneh  Finland   Helsinki      74    1.73
1     David       UK     London      78    1.75
2      John   Sweden  Stockholm      69    1.69

In [387]:
# Using function makes our code clean but you can just calculate the bmi without function
def calculate_bmi ():
    weights = df['Weight']
    heights = df['Height']
    bmi = []
    for w,h in zip(weights, heights):
        b = w/(h*h)
        bmi.append(b)
    return bmi
    
bmi = calculate_bmi()


In [388]:
df['BMI'] = bmi
df

Name  Country       City  Weight  Height        BMI
0  Asabeneh  Finland   Helsinki      74    1.73  24.725183
1     David       UK     London      78    1.75  25.469388
2      John   Sweden  Stockholm      69    1.69  24.158818

### Formating DataFrame column

The BMI of the above DataFrame has is float with many significant digits after decimal. Let's make it to have only one significant digit after point.

In [389]:
df['BMI'] = round(df['BMI'], 1)
df

Name  Country       City  Weight  Height   BMI
0  Asabeneh  Finland   Helsinki      74    1.73  24.7
1     David       UK     London      78    1.75  25.5
2      John   Sweden  Stockholm      69    1.69  24.2

The information in the DataFrame seems not yet complete, let's add birth year and current year columns.

In [390]:
birth_year = ['1769', '1985', '1990']
current_year = pd.Series(2019, index=[0, 1,2])
df['Birth Year'] = birth_year
df['Current Year'] = current_year
df

Name  Country       City  Weight  Height   BMI Birth Year  Current Year
0  Asabeneh  Finland   Helsinki      74    1.73  24.7       1769          2019
1     David       UK     London      78    1.75  25.5       1985          2019
2      John   Sweden  Stockholm      69    1.69  24.2       1990          2019

## Checking data types of Column values

In [391]:
df.Weight.dtype

dtype('int64')

In [392]:
df['Birth Year'].dtype # it give string object , we should change this to number


dtype('O')

In [393]:
df['Birth Year'] = df['Birth Year'].astype('int')
df['Birth Year'].dtype # let's check the data type now

dtype('int64')

In [394]:
df['Current Year'] = df['Current Year'].astype('int')
df['Current Year'].dtype

dtype('int64')

Now, the column values of birth year and current year are integers. We can calculate the age.

In [395]:
ages = df['Current Year'] - df['Birth Year']
ages

0    250
1     34
2     29
dtype: int64

In [396]:
df['Ages'] = ages
df

Name  Country       City  Weight  Height   BMI  Birth Year  \
0  Asabeneh  Finland   Helsinki      74    1.73  24.7        1769   
1     David       UK     London      78    1.75  25.5        1985   
2      John   Sweden  Stockholm      69    1.69  24.2        1990   

   Current Year  Ages  
0          2019   250  
1          2019    34  
2          2019    29

The person in the first row lives 250 years. It is unlikely for someone to live 250 years. Either it is a typo or the data is cooked. So lets fill that data with average of the columns without including outlier. 

mean = (34 + 29)/ 2

In [397]:
mean = (34 + 29)/ 2
mean

31.5

### Boolean Indexing

In [398]:
df[df['Ages'] > 120]

Name  Country      City  Weight  Height   BMI  Birth Year  \
0  Asabeneh  Finland  Helsinki      74    1.73  24.7        1769   

   Current Year  Ages  
0          2019   250

In [399]:
df[df['Ages'] < 120]

Name Country       City  Weight  Height   BMI  Birth Year  Current Year  \
1  David      UK     London      78    1.75  25.5        1985          2019   
2   John  Sweden  Stockholm      69    1.69  24.2        1990          2019   

   Ages  
1    34  
2    29

In [400]:
df['Ages']  = df[df['Ages'] > 120]
        
        

## Exercises: Day 25
1. Read the hacker_ness.csv file from data directory 
1. Get the first five rows
1. Get the last five rows
1. Get the title column as pandas series
1. Count the number of rows and columns
    * Filter the titles which contain python
    * Filter the titles which contain JavaScript
    * Explore the data and make sense of the data